lua/entities/gmod_wire_expression2/core/files.lua

```diff
@@ -250,108 +265,145 @@ end
 
 __e2setcost( 5 )
 
-e2function void runOnList( active )
-	run_on.list.ents[self.entity] = (active ~= 0)
+[deprecated]
+e2function void runOnList(active)
+	run_on.list.ents[self.entity] = active ~= 0
 end
 
+[deprecated]
 e2function number fileListClk()
 	return self.data.runOnFileList and 1 or 0
 end
 
+[deprecated]
 e2function number fileListClk( string dir )
 	return (self.data.runOnFileList and run_on.list.dir == dir) and 1 or 0
 end
 
 --- Hooks 'n' Shit ---
 
-registerCallback( "construct", function( self )
-	uploads[self.player] = uploads[self.player] or {
-		uploading = false,
-		uploaded = false
-	}
-	downloads[self.player] = downloads[self.player] or {
-		downloading = false,
-		downloaded = false
-	}
-	lists[self.player] = lists[self.player] or {
-		uploading = false,
-		uploaded = false
-	}
-	delays[self.player] = delays[self.player] or {
-		upload = 0,
-		download = 0,
-		list = 0
-	}
+registerCallback("construct", function(self)
+	local player = self.player
+	downloads[player] = downloads[player] or { last = {} }
+	uploads[player] = uploads[player] or { last = {} }
+	lists[player] = lists[player] or { last = {} }
 end )
 
---- Downloading ---
-util.AddNetworkString("wire_expression2_file_download_begin")
-util.AddNetworkString("wire_expression2_file_download_chunk")
-util.AddNetworkString("wire_expresison2_file_download_finish")
-timer.Create("wire_expression2_flush_file_buffer", 0.2, 0, function()
-	for ply,fdata in pairs( downloads ) do
-		if IsValid( ply ) and ply:IsPlayer() and fdata.downloading then
-			if !fdata.started then
-				net.Start("wire_expression2_file_download_begin")
-					net.WriteString(fdata.name or "")
-				net.Send(ply)
-
-				fdata.started = true
-			end
+registerCallback("destruct", function(self)
+	local player, entity = self.player, self.entity
 
-			local strlen = math.Clamp( string.len( fdata.data ), 0, download_chunk_size )
+	for k, v in ipairs(downloads[player]) do
+		if v.ent == entity then table.remove(downloads, k) end
+	end
 
-			if strlen > 0 then
-				net.Start("wire_expression2_file_download_chunk")
-				net.WriteUInt(strlen, 32)
-				net.WriteData(fdata.data, strlen)
-				net.Send(ply)
+	for k, v in ipairs(uploads[player]) do
+		if v.ent == entity then table.remove(uploads, k) end
+	end
 
-				fdata.data = string.sub( fdata.data, strlen + 1 )
-			end
+	for k, v in ipairs(lists[player]) do
+		if v.ent == entity then table.remove(lists, k) end
+	end
```

Comment:
This is worrying, what happens if the chip is removed during an upload/download? Would another queued chip receive the correct data?

Also these need to be while loops because you are modifying the array that you are iterating.

Reply:
The next chip in the queue would receive a file transfer error because the flag `uploading` would be unset, but I'll look into avoiding that.

Reply:
This has been mitigated in recent commits.